File: ribosome.rb
#
# Copyright (c) 2014 Martin Sustrik  All rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom
# the Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
#

################################################################################
#  DNA helper functions.                                                       #
################################################################################

# Print out the error and terminate the generation.
def dnaerror(s)
    $stderr.write("#{$dnafile}:#{$ln} - #{s}\n")
    exit()
end

# Remove whitespace from the beginning of the . or + line.
# Returns trimmed line and the amount of whitespace removed.
def ltrim(s)

    # Find out the amount of whitespace at the beginning of the line.
    ws = 0;
    for i in 1..s.size()
        if(s[i] == ?\s)
            ws += 1
            next
        end
        if(s[i] == ?\t)
            ws += 8
        end
        break
    end

    # Cut off the control character and any subsequent whitespace.
    s = s[i..-2]

    return ws,s
end


################################################################################
#  RNA helper functions.                                                       #
################################################################################

rnahelpers = '

module Ribosome

    def Ribosome.openroot(name)
        if(name[-4..-1] == ".xml")
            require "rexml/document"
            return REXML::Document.new(File.new(name)).root
        elsif(name[-5..-1] == ".json")
            require "rubygems"
            require "json"
            return JSON.parse(File.read(name))
        else
            $stderr.puts("input file must be either .json or .xml")
            exit()
        end
    end

    def Ribosome.write(s)
        $stack.last.push(s)
    end

    def Ribosome.close()
        for l in $stack.last
            $out.write(l)
        end
        $stack = [[]]
        $out.write("\n");
        if ($outisafile)
            $out.close()
        end
    end

    def Ribosome.expand(s, b)

        # Find all occurences of @{.
        i = -1
        while true
            i = s.index(\'@{\', i + 1)
            if(i == nil)
                break;
            end
            j = i + 1

            # Find corresponding }.
            par = 0;
            while true
                if(s[j] == ?{)
                    par += 1
                end
                if(s[j] == ?})
                    par -= 1
                end
                if(par == 0)
                    break
                end
                j += 1
            end

            # Replace the expression with its value.
            expr = s[i + 2..j - 1]
            $stack.push([])
            val = eval(expr, b)
            top = $stack.pop()
            if(top.empty?)

                # Classic functions.
                s[i..j] = val.to_s()
                i += val.to_s().size
            else

                # Ribosome functions.
                res = top.join()
                s[i..j] = res
                i += res.size
            end
            
        end

        return s
    end

    def Ribosome.adjust(line, ws)

        a = line.to_a()

        # Find top, bottom and left boundary of the block of text.
        i = 0
        top = -1
        bottom = -1
        left = -1
        for l in a
            if(!l.lstrip().empty?)
                if(top == -1)
                    top = i;
                end
                bottom = i;
                if (left == -1)
                    left = l.size() - l.lstrip().size()
                else
                    left = [left, l.size() - l.lstrip().size()].min
                end
            end
            i += 1
        end

        # Strip the top and bottom whitespace.
        if(top == -1)
            a = []
        else
            a = a[top..bottom]
        end

        # Strip the original whitespace from the left.
        # Add the amount of whitespace specified in the argument.
        for i in 0..a.size() - 1
            a[i] = " " * ws + a[i][left..-1]
        end

        # Return the adjusted block. Strip off the trailing newline.
        res = a.join
        if(res[-1] == ?\n)
            res = res[0..-2]
        end
        return res;

    end

    def Ribosome.dot(line, ws, bind)
        write("\n")
        write(adjust(expand(line, bind), ws))
    end

    def Ribosome.plus(line, bind)
        write(expand(line, bind))
    end

    # Initialise the ribosome stack. Each level on the stack contains a block of
    # text in the form of array of lines (strings).
    $stack = [[]]

    # Initialise output channel.
    $out = $stdout
    $outisafile = false

end

'

################################################################################
#  Main function.                                                              #
################################################################################

# Parse the command line arguments.
if(ARGV.size() != 1 && ARGV.size() != 2)
    puts("usage: ribosome <dna-file> [<input-file>]")
    exit()
end
$dnafile = ARGV[0]
if (ARGV.size < 2)
    infile = ""
else
    infile = ARGV[1]
end

# Open the files for the DNA-to-RNA translation step.
$ln = 0
if($dnafile[-4..-1] == ".dna")
    rnafile = $dnafile[0..-5] + ".rna.rb"
else
    rnafile = $dnafile + ".rna.rb"
end
dna = File.open($dnafile, "r")
rna = File.open(rnafile, "w")

# Add RNA helper functions.

rna.write(rnahelpers)

# Open the input file.
rna.write("if(ARGV.size > 0)\n")
rna.write("    $root = Ribosome.openroot(ARGV[0])\n")
rna.write("end\n\n")

# Process the DNA file, line-by-line.
while(line = dna.gets())

    # We are counting lines so that we can report line numbers in errors.
    $ln += 1

    # Empty lines are ignored.
    if(line.size() == 0 || line[0] == ?\n)
        next
    end

    # Lines starting with '!' are ribosome commands.
    if(line[0] == ?!)

        # Parse the arguments. This should probably be done in a more
        # sophisticated way in the future to allow arguments containing spaces.
        words = line[1..-1].split(/\s+/)

        # !separate is used to insert separators between
        # the iterations of a loop.
        #if(words[0] == 'separate') {
        #    rna.write("____first_#{$ln}____ = true\n")
        #    line = dna.gets()
        #    $ln += 1
        #    next
        #}

        # Remaining commands can be used only in the outermost scope.
        rna.write("if(Ribosome.$stack.size > 1)\n")
        rna.write("    $stderr.write(\"#{$dnafile}:#{$ln} - command '#{words[0]}' used in a nested function\\n\")\n")
        rna.write("    exit()\n")
        rna.write("end\n")

        # !output redirects the output to a different file.
        if(words[0] == 'output')
            if(words.size() != 2)
                dnaerror("command 'output' expects one argument")
            end
            rna.write("Ribosome.close()\n")
            rna.write("Ribosome.$outisafile = true\n")
            rna.write("Ribosome.$out = File.open(#{words[1]}, 'w')\n")
            next
        end

        # !stdout redirects the output to stdout.
        if(words[0] == 'stdout')
            if(words.size() != 1)
                dnaerror("command 'stdout' expects no arguments")
            end
            rna.write("Ribosome.close()\n")
            rna.write("Ribosome.$outisafile = false\n")
            rna.write("Ribosome.$out = $stdout\n")
            next
        end

        # Invalid command.
        $stderr.puts("#{$dnafile}:#{$ln} - invalid command '#{words[0]}'")
        exit()
    end

    # Dot indicates thar the following text should be copied to a new line
    # in the output. The text is properly indented.
    if(line[0] == ?.)
        ws,line = ltrim(line)
        rna.write("Ribosome.dot(#{line.inspect()}, #{ws}, binding)\n")
        next
    end

    # Plus sign (+) indicates that the following text should be copied to
    # the output without moving to a new line. No indenting kung-fu is done. 
    if(line[0] == ?+)
        ws,line = ltrim(line)
        rna.write("Ribosome.plus(#{line.inspect()}, binding)\n")
        next
    end

    # All other lines are copied to the RNA file verbatim.
    rna.write(line)

end

# Flush the output file.
rna.write("Ribosome.close()\n\n")

# Flush the RNA file.
rna.close()
dna.close()

# Execute the RNA file.
system("ruby #{rnafile} #{infile}")

# Delete the RNA file.
# For now we are letting the file be to help with debugging of ribosome.
# File.delete(rnafile)

File: ribosome.py
#!/usr/bin/env python

from __future__ import print_function, division

#
# Copyright (c) 2015 Ali Zaidi  All rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom
# the Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
#

import os
import re
import sys
import ast, _ast
import inspect
import unicodedata

line = lambda level=1: inspect.stack()[level][2]
templ = lambda s, level=1: s.format(**inspect.stack()[level][0].f_locals)
dnacontext = lambda: inspect.currentframe().f_back.f_back.f_locals
contextvar = lambda s: dnacontext[s]
tabcollate = lambda s, ts: re.sub('^( {'+str(ts)+'})*', lambda m: '\t'*(m.end()//ts), s)


class Document:
    """ Text document that supports multi-line expansion in append functions.

    All public manipulation functions accept a list of line chunks that are stuck to each other like this:
      [-----] [-chunk 2-] [---chunk 3----] [-chunk-]
      |chunk| [---------]                  |   4   |
      |  1  |                              [-------]
      [-----]
    Each chunk may by itself contain multiple lines.
    """

    def __init__(self, out=sys.stdout, s='', tabsize=0):
        self.text = []
        self.cur = s.splitlines()
        self.out = out
        self.tabsize = tabsize

    def _add_elem(self, *elines):
        w = max(len(l) for l in self.cur or [''])
        d = len(elines)-len(self.cur) # to pad cur and line to equal lengths
        self.cur = [c.ljust(w)+l for c,l in zip(self.cur+['']*max(d, 0), elines+('',)*max(-d, 0))]

    def add(self, *line):
        """ Weld the supplied block to the right of the current block """
        for elem in line:
            self._add_elem(*(elem.splitlines() or ['']))

    def dot(self, *line):
        """ Weld the supplied block to the bottom of the current block """
        self.text += self.cur
        self.cur = []
        self.add(*line)

    def align(self, *line):
        """ Like dot, but left-align content with previous line """
        indent = ' '*max(re.match('[\s]*', l).end() for l in self.cur)
        self.dot(*( indent+l for l in line ))

    # Output control
    def write(self):
        self.dot()
        self.out.write('\n'.join(tabcollate(line, self.tabsize) if self.tabsize else line for line in self.text))

    def close(self):
        self.write()
        self.out.write('\n')
        self.out.close()


_separate_state = {}
def separate(sep, sid, add):
    global _separate_state
    if not _separate_state.get(sid):
        _separate_state[sid] = True
    else:
        add(sep)

            
def include(file_or_name, _globals=None):
    """ Include another DNA document

    This document will have access to any local variables and functions from the including document. The including
    document, in turn, will **NOT** have access to anything locally declared in the included document.
    """
    includefile = open(file_or_name, 'r') if isinstance(file_or_name, str) else file_or_name

    caller_f = inspect.currentframe().f_back
    _globals = _globals or caller_f.f_globals.copy()
    filename, warnctx = _globals['_filename'], _globals['_warnctx']
    lineno = caller_f.f_lineno
    def newwarnctx():
        warnctx(s)
        print('At {}:{}'.format(filename, lineno), file=sys.stderr)

    _globals.update({ '_filename': includefile.name, '_warnctx': newwarnctx })

    with includefile as f:
        tree = parse_lines(includefile.name, includefile.readlines(), newwarnctx)
        code = compile(tree, includefile.name, 'exec')
        exec(code, _globals)

def parse_lines(filename, lines, warnctx):
    def warn(s):
        warnctx()
        print('From {}:{}'.format(__file__, line(2)), file=sys.stderr)
        print(templ(s, 2), file=sys.stderr)
        print('> '+rawline, file=sys.stderr)

    code = []
    for lineno,rawline in enumerate(lines):
        indent,     dot,     shortcmd,    line,    rspace,     dollar = re.match(
        r'^([\s]*)'+r'(\.?)'+r'(/[+=!])?'+r'(.*?)'+r'([\s]*?)'+r'(\$?)\n?$', rawline).groups()

        if rspace and not dollar:
            warn('Trailing space in line not dollar-terminated')

        if '\t' in rawline:
            warn("Line contains tabs. For consistent results, consider using spaces.")

        if not dot:
            code.append(rawline.rstrip('\n'))
            continue

        def repl(s):
            m = re.search(r'([@&])0?([1-9]?)\{', s)
            if not m:
                return repr(s)
            start,end, = m.span()
            op, ncount = m.groups()
            lvl = 1
            for b in re.finditer('[{}]', s[end:]):
                lvl += 1 if b.group(0) == '{' else -1
                if lvl == 0:
                    if ncount:
                        if ncount != '1':
                            s = s[:m.end()-2] + str(int(ncount)-1) + s[m.end()-1:]
                        else:
                            s = s[:m.end()-2] + s[m.end()-1:]
                        return repr(s[:end+b.end()]) + repl(s[end+b.end():])
                    else:
                        return ', '.join([ repr(s[:start]),
                            ("str({}).strip()" if op == '@' else "str({})").format(s[end:end+b.start()])
                            , repl(s[end+b.end():]) ])

        if not shortcmd:
            cmd, args = 'dot', repl(line)
        elif shortcmd == '/+': # Append the line to the previous line.
            cmd, args = 'add', repl(line)
        elif shortcmd == '/=': # /= Align the line with the previous line.
            cmd, args = 'align', repl(line)
        else:
            m = re.match(r'(^[0-9A-Za-z_]+)\((.*)\)$', line)
            if not m:
                warn('Invalid command syntax. Expected: /!command(arg, etc)')
                code.append('# Invalid command syntax: {}'.format(rawline))
                continue
            cmd, args = m.groups()
        code.append(templ('{indent}{cmd}({args})'))

    if DEBUG:
        print('GENERATED:')
        print('\n'.join(code))
        print('---')
    tree = ast.parse('\n'.join(code), filename)

    # Fix up generated AST for "separate" commands
    # from _ast import * does not work in python2
    done = []
    for node in ast.walk(tree):
        for val in node.__dict__.values():
            if type(val) is list:
                indices = [i for i,e in enumerate(val) if
                        type(e) is _ast.Expr and
                        type(e.value) is _ast.Call and
                        type(e.value.func) is _ast.Name and
                        e.value.func.id == 'separate']
                for idx in reversed(indices):
                    sepcall, loop = val[idx], val[idx+1]
                    if sepcall in done:
                        continue #FIXME dirty hack
                    if type(loop) not in (_ast.For, _ast.While, _ast.FunctionDef):
                        warn('"separate" command must be followed by loop or function definition.')
                        continue
                    sepcall.value.args.append(ast.Str('{}:{}'.format(filename, idx)))
                    sepcall.value.args.append(ast.Name('add', ast.Load()))
                    ast.fix_missing_locations(sepcall)
                    loop.body = [sepcall] + loop.body
                    done.append(sepcall)
                    del val[idx]
    return tree

# Escape sequences
at    = '@'
amp   = '&'
slash = '/'

def runfile(f):
    _doc = Document()

    def redirect(f):
        _doc.close()
        _doc = Document(f)

    def tabsize(ts):
        _doc.tabsize = ts

    _globals = globals().copy()
    _globals.update({
        '_filename': '<ribosome>',
        '_warnctx' : lambda: None,
        '_doc'     : _doc,
        'redirect' : redirect,
        'tabsize'  : tabsize,
        'add'      : lambda *line: _doc.add(*line),
        'dot'      : lambda *line: _doc.dot(*line),
        'align'    : lambda *line: _doc.align(*line),
        'close'    : lambda: _doc.close(),
        'stdout'   : lambda: redirect(sys.stdout),
        'stderr'   : lambda: redirect(sys.stderr),
        'output'   : lambda filename: redirect(open(filename, "w")),
        'append'   : lambda filename: redirect(open(filename, "a")) })


    include(f, _globals)
    _doc.close()

DEBUG = False
if __name__ == '__main__':
    # Set up the arguments parser.
    import argparse
    parser = argparse.ArgumentParser(prog="ribosome code generator, version 1.16")
    parser.add_argument('dna', type=argparse.FileType('r'), default=sys.stdin)
    parser.add_argument('--debug', action='store_true')
    args = parser.parse_args()
    DEBUG = args.debug
    runfile(args.dna)

